openc3/lib/openc3/microservices/reducer_microservice.rb

```diff
@@ -155,50 +170,78 @@ def shutdown
+        sample_name = name + '_sample_seconds'
+        @metric.set(name: sample_name, value: elapsed, type: 'gauge', unit: 'seconds')
+        max_name = name + '_max_seconds'
+        previous_max = @previous_metrics[max_name] || 0.0
+        if elapsed > previous_max
+          @metric.set(name: max_name, value: elapsed, type: 'gauge', unit: 'seconds')
+          @previous_metrics[max_name] = elapsed
+        end
+      end
     end
 
     def reduce_minute
-      metric(MINUTE_METRIC) do
-        ReducerModel
-          .all_files(type: :DECOM, target: @target_name, scope: @scope)
-          .each do |file|
-            process_file(file, 'minute', MINUTE_ENTRY_NSECS, MINUTE_FILE_NSECS)
-            ReducerModel.rm_file(file)
+      @mutex.synchronize do
+        metric(MINUTE_METRIC) do
+          processed = false
+          ReducerModel
+            .all_files(type: :DECOM, target: @target_name, scope: @scope)
+            .each do |file|
+              process_file(file, 'minute', MINUTE_ENTRY_NSECS, MINUTE_FILE_NSECS)
+              ReducerModel.rm_file(file)
+              processed = true
           end
+          processed # return to yield
+        end
       end
     end
 
     def reduce_hour
-      metric(HOUR_METRIC) do
-        ReducerModel
-          .all_files(type: :MINUTE, target: @target_name, scope: @scope)
-          .each do |file|
-            process_file(file, 'hour', HOUR_ENTRY_NSECS, HOUR_FILE_NSECS)
-            ReducerModel.rm_file(file)
+      @mutex.synchronize do
+        metric(HOUR_METRIC) do
+          processed = false
+          ReducerModel
+            .all_files(type: :MINUTE, target: @target_name, scope: @scope)
+            .each do |file|
+              process_file(file, 'hour', HOUR_ENTRY_NSECS, HOUR_FILE_NSECS)
+              ReducerModel.rm_file(file)
+              processed = true
           end
+          processed # return to yield
+        end
       end
     end
 
     def reduce_day
-      metric(DAY_METRIC) do
-        ReducerModel
-          .all_files(type: :HOUR, target: @target_name, scope: @scope)
-          .each do |file|
-            process_file(file, 'day', DAY_ENTRY_NSECS, DAY_FILE_NSECS)
-            ReducerModel.rm_file(file)
+      @mutex.synchronize do
+        metric(DAY_METRIC) do
+          processed = false
+          ReducerModel
+            .all_files(type: :HOUR, target: @target_name, scope: @scope)
+            .each do |file|
+              process_file(file, 'day', DAY_ENTRY_NSECS, DAY_FILE_NSECS)
+              ReducerModel.rm_file(file)
+              processed = true
           end
+          processed # return to yield
+        end
       end
     end
 
     def process_file(filename, type, entry_nanoseconds, file_nanoseconds)
       throttle = OpenC3::Throttle.new(@max_cpu_utilization)
       file = BucketFile.new(filename)
       file.retrieve
-      unless file.local_path
-        @logger.warn("Reducer Warning: #{filename}: Could not be retrieved")
```

Comment:
Previously we just warned and then carried on? That makes no sense.

Reply:
Yeah - Not correct before